master/internal/db/postgres_groups.go

```diff
@@ -0,0 +1,194 @@
+// successful search.
+func (db *PgDB) SearchGroups(ctx context.Context,
+	userBelongsTo model.UserID) ([]model.Group, error) {
+	var groups []model.Group
+	query := Bun().NewSelect().Model(&groups).Distinct()
+
+	if userBelongsTo > 0 {
+		query = query.
+			Join("INNER JOIN user_group_membership AS ugm ON ugm.group_id=groups.id").
+			Where("ugm.user_id = ?", userBelongsTo)
+	}
+
+	err := query.Scan(ctx)
+
+	return groups, err
+}
+
+// DeleteGroup deletes a group from the database. Returns ErrNotFound if the
+// group doesn't exist.
+func (db *PgDB) DeleteGroup(ctx context.Context, gid int) error {
+	tx, err := Bun().BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	_, err = tx.NewDelete().
+		Table("user_group_membership").
+		Where("group_id = ?", gid).
+		Exec(ctx)
+	if err != nil {
+		return err
+	}
+
+	res, err := tx.NewDelete().Model(&model.Group{ID: gid}).WherePK().Exec(ctx)
+	if foundErr := mustHaveAffectedRows(res, err); foundErr != nil {
+		return matchSentinelError(foundErr)
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return matchSentinelError(err)
+	}
+
+	return nil
+}
+
+// UpdateGroup updates a group in the database. Returns ErrNotFound if the
+// group isn't found.
+func (db *PgDB) UpdateGroup(ctx context.Context, group model.Group) error {
+	res, err := Bun().NewUpdate().Model(&group).WherePK().Exec(ctx)
+
+	return matchSentinelError(mustHaveAffectedRows(res, err))
+}
+
+// AddUsersToGroup adds users to a group by creating GroupMembership rows.
+// Returns ErrNotFound if the group isn't found or ErrDuplicateRow if one
+// of the users is already in the group.
+func (db *PgDB) AddUsersToGroup(ctx context.Context, gid int, uids ...model.UserID) error {
+	if len(uids) < 1 {
+		return nil
+	}
+
+	groupMem := make([]model.GroupMembership, 0, len(uids))
+	for _, uid := range uids {
+		groupMem = append(groupMem, model.GroupMembership{
+			UserID:  uid,
+			GroupID: gid,
+		})
+	}
+
+	tx, err := Bun().BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	res, err := tx.NewInsert().Model(&groupMem).Exec(ctx)
+	if foundErr := mustHaveAffectedRows(res, err); foundErr != nil {
+		return matchSentinelError(foundErr)
+	}
+
+	err = tx.Commit()
+	return matchSentinelError(err)
+}
+
+// RemoveUsersFromGroup removes users from a group. Removes nothing and
+// returns ErrNotFound if the group or one of the users' membership rows
+// aren't found.
+func (db *PgDB) RemoveUsersFromGroup(ctx context.Context, gid int, uids ...model.UserID) error {
+	if len(uids) < 1 {
+		return nil
+	}
+
+	tx, err := Bun().BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	res, err := tx.NewDelete().Table("user_group_membership").
+		Where("group_id = ?", gid).
+		Where("user_id IN (?)", bun.In(uids)).
+		Exec(ctx)
+	if foundErr := mustHaveAffectedRows(res, err); foundErr != nil {
```

Comment:
I believe this won't throw an error if I call it with a user in the group and a user not in the group since we are only checking if ```affectedRows``` is nonzero compared to the function comment above implying we need to check if ```affectedRows``` is ```len(uids)```? 

Adding users to groups doesn't have this issue because it protected by the foreign key / duplicate key pg I think?

Reply:
You're absolutely right about both. Time to add to the test cases.